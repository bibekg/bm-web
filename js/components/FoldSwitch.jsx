// @flow

import * as React from 'react'
import styled from 'styled-components'
import { ExpandMoreIcon } from 'components/icons'
import { Text } from 'components/typography'
import { colors } from 'styles'

type PropsType = {
  isFolded: boolean,
  text?: string,
  onClick: (SyntheticEvent<*>) => void,
  iconColor: string
}

const FoldSwitchDiv = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
`

const RotateDiv = styled.div`
  transform: rotate(${props => (props.isFolded ? '0' : '180')}deg);
  transform-origin: center;
  transition: 0.2s;
`

const FoldedText = Text.extend`
  margin: 0;
  text-align: center;

  /* TODO: add a fade out animation to this text */
`

function FoldSwitch(props: PropsType): React.Element<*> {
  return (
    <FoldSwitchDiv onClick={props.onClick}>
      {props.isFolded && props.text && <FoldedText>{props.text}</FoldedText>}
      <RotateDiv isFolded={props.isFolded}>
        <ExpandMoreIcon fill={props.iconColor} size={40} />
      </RotateDiv>
    </FoldSwitchDiv>
  )
}

FoldSwitch.defaultProps = {
  iconColor: colors.lightBlue
}

export default FoldSwitch
