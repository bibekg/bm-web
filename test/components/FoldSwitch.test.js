import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import 'jest-styled-components'

import FoldSwitch from 'components/FoldSwitch'

describe('<FoldSwitch />', () => {
  test('should render correctly with no options', () => {
    const wrapper = shallow(<FoldSwitch />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render correctly folded with options', () => {
    const wrapper = shallow(<FoldSwitch isFolded text="Test text" iconColor="#123456" />)
    expect(wrapper).toMatchSnapshot()

    // check that text is present and correct
    expect(
      wrapper
        .childAt(0) // AnimateHeight
        .childAt(0) // styled.p
        .children()
        .text()
    ).toEqual('Test text')
    expect(wrapper.childAt(0).prop('height')).toEqual('auto')

    // check that div correctly received isFolded prop
    expect(wrapper.childAt(1).prop('isFolded')).toEqual(true)

    // check that fill color is correct
    expect(wrapper.find('ExpandMoreIcon').prop('fill')).toEqual('#123456')

    // check that arrow rotation is correct
    expect(wrapper.childAt(1)).toHaveStyleRule('transform', 'rotate(0deg)')
  })

  test('should unfold and unrender text when isFolded set to false', () => {
    const wrapper = shallow(<FoldSwitch isFolded text="Test text" iconColor="#123456" />)
    wrapper.setProps({ isFolded: false })
    expect(wrapper).toMatchSnapshot()

    // check that text is not visible
    expect(wrapper.childAt(0).prop('height')).toEqual(0)

    // check that div correctly received isFolded prop
    expect(wrapper.childAt(1).prop('isFolded')).toEqual(false)

    // check that arrow rotation is correct
    expect(wrapper.childAt(1)).toHaveStyleRule('transform', 'rotate(180deg)')
  })

  test('should refold and rerender text when isFolded set to true', () => {
    const wrapper = shallow(<FoldSwitch text="Test text" iconColor="#123456" />)
    wrapper.setProps({ isFolded: true })
    expect(wrapper).toMatchSnapshot()

    // check that text exists again
    expect(
      wrapper
        .childAt(0)
        .childAt(0)
        .children()
        .text()
    ).toEqual('Test text')
    expect(wrapper.childAt(0).prop('height')).toEqual('auto')

    // check that div correctly received isFolded prop
    expect(wrapper.childAt(1).prop('isFolded')).toEqual(true)

    // check that arrow rotation is correct
    expect(wrapper.childAt(1)).toHaveStyleRule('transform', 'rotate(0deg)')
  })

  test('should handle onClick correctly', () => {
    const clickFcn = jest.fn()
    const wrapper = shallow(<FoldSwitch text="Test text" iconColor="#123456" onClick={clickFcn} />)

    wrapper.simulate('click')
    expect(clickFcn).toHaveBeenCalled()
  })
})
