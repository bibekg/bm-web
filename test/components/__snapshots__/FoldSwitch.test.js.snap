// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FoldSwitch /> should refold and rerender text when isFolded set to true 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.p
    isFolded={true}
  >
    Test text
  </styled.p>
  <styled.div
    isFolded={true}
  >
    <ExpandMoreIcon
      fill="#123456"
      size={40}
    />
  </styled.div>
</styled.div>
`;

exports[`<FoldSwitch /> should render correctly folded with options 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.p
    isFolded={true}
  >
    Test text
  </styled.p>
  <styled.div
    isFolded={true}
  >
    <ExpandMoreIcon
      fill="#123456"
      size={40}
    />
  </styled.div>
</styled.div>
`;

exports[`<FoldSwitch /> should render correctly with no options 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isFolded={false}
  >
    <ExpandMoreIcon
      fill="rgba(162, 198, 248, 1)"
      size={40}
    />
  </styled.div>
</styled.div>
`;

exports[`<FoldSwitch /> should unfold and unrender text when isFolded set to false 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.p
    isFolded={false}
  >
    Test text
  </styled.p>
  <styled.div
    isFolded={false}
  >
    <ExpandMoreIcon
      fill="#123456"
      size={40}
    />
  </styled.div>
</styled.div>
`;
